internal/provider/resource_workspace_current_state.go

```diff
@@ -0,0 +1,313 @@
+				MarkdownDescription: "Whether to teardown the deployment of the module to the workspace.",
+				Description:         "Whether to teardown the deployment of the module to the workspace.",
+				Optional:            true,
+				Computed:            true, // API sets a default value of false if not specified.
+				// Can be updated in place, so no RequiresReplace plan modifier.
+			},
+		},
+	}
+}
+
+// Configure lets the provider implement the ResourceWithConfigure interface.
+func (t *workspaceCurrentStateResource) Configure(_ context.Context,
+	req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *workspaceCurrentStateResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from workspace current state.
+	var workspaceCurrentState WorkspaceCurrentStateModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &workspaceCurrentState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If teardown is not specified, make it false.
+	workspaceCurrentState.setDefaultTeardown()
+
+	if !workspaceCurrentState.Teardown.ValueBool() {
+		t.doApplyOrDestroyRun(ctx, workspaceCurrentState, resp.Diagnostics)
+	}
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, workspaceCurrentState)...)
+}
+
+func (t *workspaceCurrentStateResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state WorkspaceCurrentStateModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// No need to set Teardown's default value.
+
+	// There is no model in the Tharsis API, so there's nothing really to read.
+
+	// Set the refreshed state, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (t *workspaceCurrentStateResource) Update(ctx context.Context,
+	req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	// Retrieve values from plan.
+	var plan WorkspaceCurrentStateModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If teardown is not specified, make it false.
+	plan.setDefaultTeardown()
+
+	// Apply or destroy, depending on the Teardown flag.
+	t.doApplyOrDestroyRun(ctx, plan, resp.Diagnostics)
+
+	// Set the response state to the fully-populated plan, with or without error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+func (t *workspaceCurrentStateResource) Delete(ctx context.Context,
+	req resource.DeleteRequest, resp *resource.DeleteResponse) {
+
+	// Get the current state.
+	var state WorkspaceCurrentStateModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
```

Comment:
Delete should only create a destroy run if the currently deployed module matches the module and version specified in the current_state_version. It should return an error if a different module is currently deployed.